### DISCUSSION TITLE
Reasoning behind current branch workflow?

### DISCUSSION THREAD

**Discussion by sugan-reden** [Epicrex/3DArtistsHandbook#9] [Q&A] [answered]
Hi Sidney,

I was just wondering what was the reasoning behind the current branch structure (i.e. having the main content of the site on the v4 branch instead of the main)?

Would you perhaps be open to swapping them?

Hope you're doing well,
Cheers 🍻

**Comment by sidney-eliot** [comment]
Hey Sugan. (Or what would you like me to call you.) Had a busy week, but I have time today and this weekend. I'm still going to reply to the other stuff you wrote, but will start with this.

I did this so that the handbook is more approachable by less tech experienced people that might be scared of by being greeted by a long list of code files and such. If GitHub offered the option to put the README file above the file structure, I might not have done that.

A second reason is that the backend structure on which the handbook is based on might change over the years, like a new quartz version (a major version that can't just be fetched but requires a new branch. Like for example, the jump from quartz 3 to quartz 4 which changed the entire framework from Hugo to Node JS). But who know, maybe in the future there are better options than quartz (not that I'm trying to get away from quartz or anything, I just like to stay open about stuff like that). All that really matters about the handbook is the markdown pages, which are very universal. If one were to change to a new framework, for example, which maybe handles internal page linking differently for example, then one could just simply let some script run through all the markdown pages to update the internal linking method from [[#exmaple|Example]] to who knows what.

This flexibility is also one of the reason why I love quartz so much over a more conventional way of making websites via HTML files.

If you still think this makes sense for whatever reason, then tell me about it. But the way I see it currently is that it makes sense to have one branch as the so to say "Lobby Branch" which is certain to not change as it simply contains the handbook description and maybe other small resources like 3D software preference files and so on.

What I was also thinking about was to perhaps move the CONTRIBUTING.md as well as the CODE_OF_CONDUCT.md from the v4 branch to the main branch so that they are more easily found. The problem with that is that then they wouldn't be in the v4 anymore, so if someone clones that branch and starts working, they won't see the CONTRIBUTING.md. Or one updates the files in both branches, which seems like not the best idea to me.

I guess this is why most repos have the development branch as the default branch.

As you can see, this is something I'm kind of split about. I also just updated the README to make this structure clearer, linking to the v4 branch directly at the top as well.

Cheers 🦑

**Comment by sidney-eliot** [comment]
I actually would find the perfect structure to be a branch called main containing; like it does now the README, and resources like software preference files, in addition to the aforementioned CONTRIBUTING.md / CODE_OF_CONDUCT.md file (and here comes the punchline), a folder called pages containing the markdown files and a folder called images. The second branch would then be called backend and would contain the quartz Node.js. This would nicely separate the content from the website architecture.

The deployment action could then somehow deploy the backend and fetch the pages and images from the main branch.

The first problem I can think of, however, with this approach, is that building the handbook locally would be a lot more tricky. One would have to modify the Node.js serve command.

What I had once also thought about in the past (although I see this comes with far too many downsides) is to have the handbook only contain the content files and have a separate repository that deploys the handbook, fetching the content pages from the handbook repo.

**Comment by sugan-reden** [comment]
Hi Sidney,

Thought I'd have time yesterday but got visits that took more time than expected ^^. That said, I have time now :)

> ... (Or what would you like me to call you.) ...

Sugan is fine here, or anywhere else :)

> Had a busy week, but I have time today and this weekend. I'm still going to reply to the other stuff you wrote, but will start with this.

Relatable man, for real holidays do being crazy right now.

---

> I did this so that the handbook is more approachable by less tech experienced people that might be scared of by being greeted by a long list of code files and such. If GitHub offered the option to put the README file above the file structure, I might not have done that.

This is a good point to bring up, and accessibility is important for websites, especially for guides/books. It is also a gripe I had with GitHub when I first started using it, but I'm more used to it now.
However, this makes wonder: 
- What kind of traffic do you mainly expect to get on GitHub? 
- What are your main expectations of the type of audience that will fall onto your site vs. repo? 
- If you expect the handbook (Quartz) to be the first thing they they find, is this level of accessibility truly necessary?

To me, if the user is even only somewhat tech-savvy (or at the very least curious), seeing the structure directly -- I assume -- will probably make it easier for them to understand the structure of the project directly, files and backend both. I expect most users that find the repo to somewhat knowledgeable about this. 
However, if the user is still uncomfortable with git or using using version control in general, or fiddling with editing by pulling/pushing from local machine, then they can simply use web-based editor without the need to install anything. The main reason I think `main` should be the docs is this, since less git-experienced users will ultimately have less clicks to go through to edit, or in other words:

_`repo page` > `fork button` > `go to their own fork` > `go to 'docs/' in file structure` > `find file they want to edit` > `✏️ Edit this page button` > `add description and commit/push changes`_

instead of currently:

_`repo page` > `fork button` > `switch to v4 branch` > ... (idem as above) > `add desciption and ensure they commit/push **to the right branch**`_

In my opinion, I see the main traffic model as the following:

```mermaid
flowchart TD;

    A@{ shape: rounded, label: "User finds website..."};
    B@{ shape: decision, label: "User is tech-savvy"};
    C@{ shape: event, label: "Yes"};
    D@{ shape: event, label: "No"};
    E@{ shape: subproc, label: "through GitHub"};
    F@{ shape: subproc, label: "through Quartz" };
    G@{ shape: subproc, label: "through other means, e.g. X, Discord, ArtStation, etc."};
    H@{ shape: event, label: "User wants to contribute"};
    I@{
        shape: proc
        label: "User can contribute however they like or already know but probably by forking, cloning, and editing on local machine and the text editor of their choice and pushing changes (with PRs)."
    }
    J@{
        shape: proc
        label: "User is invited to contribute by forking and editing with the github.dev web-based editor for more accessibility, then pushing (with PRs) the changes from their fork on github.com"
    }


  
    A --> E & F & G;
    E & F & G --> H;
    H --> B;
    B --> C & D;
    C --> I;
    D --> J;
```

Please tell me how you see it ^^.

---

> A second reason is that the backend structure on which the handbook is based on might change over the years, like a new quartz version (a major version that can't just be fetched but requires a new branch. Like for example, the jump from quartz 3 to quartz 4 which changed the entire framework from Hugo to Node JS).

That's an interesting point to bring up, I always assumed that, while making a major migration like this, if there was ever trouble that came up, we could always revert since we're using version control (git). For more peace of of mind, we could always start the migration process on a different branch and then merge (i.e. branch out > upgrade quartz > migrate and test out new structure while in the branch > if all works well then merge, if not then downgrade quartz and delete branch). Curious to know what you think of this?

> But who know, maybe in the future there are better options than quartz (not that I'm trying to get away from quartz or anything, I just like to stay open about stuff like that). All that really matters about the handbook is the markdown pages, which are very universal. If one were to change to a new framework, for example, which maybe handles internal page linking differently for example, then one could just simply let some script run through all the markdown pages to update the internal linking method from [[#exmaple|Example]] to who knows what.
>
>This flexibility is also one of the reason why I love quartz so much over a more conventional way of making websites via HTML files.

I'm curious about you mean more specifically by 'more conventional way of making websites'.
Are you talking about stacks like JS + CSS + HTML? Because there a lot of ways of making websites, this stack being only one, although popular.

I'm curious to know what you think of SSGs (Static Site Generators), do you know of them?
If not, would you be interested in checking out these options?
- [mystmd](https://mystmd.org/guide) or [main page with small playground](https://mystmd.org/)
- [VuePress](https://vuepress.vuejs.org/) with a theme like [Hope](https://theme-hope.vuejs.press/)
- [Sphinx](https://www.sphinx-doc.org/en/master/#) with the [Shibuya](https://shibuya.lepture.com/) or [Book Theme](https://sphinx-book-theme.readthedocs.io/en/stable/index.html)
- [Astro](https://astro.build/)'s [Starlight](https://starlight.astro.build/)

These are some options of the top of my head for documentation or docs-like websites (books). Some of these include an export to PDF, if that would interest you.
The other thing to consider though is that most of these do not have a graph-view like Obsidian. 
They also diverge from it a bit by not necessarily being linked to your notes, I assumed this point was more important, but I'd actually be curious to know your thoughts on this.
(Although they potentially could be, but you may find yourself with two conflicting parser/syntax on the same .md files.)

---

> If you still think this makes sense for whatever reason, then tell me about it. But the way I see it currently is that it makes sense to have one branch as the so to say "Lobby Branch" which is certain to not change as it simply contains the handbook description and maybe other small resources like 3D software preference files and so on.

This does make sense too, as it is certainly not the end of the world for developers or tech-savvy people, and does make it less intimidating for non-technical users.
That said, to go back on my first section for a bit, if the user is expected to find the handbook (Quartz) first, then it may be advantageous to put a *Contributing* section inside of the website, describing the *How-to's* to non-technical and more technical both. (The more technical section could, for example, simply lead to the `CONTRIBUTING.md`.

---

> What I was also thinking about was to perhaps move the CONTRIBUTING.md as well as the CODE_OF_CONDUCT.md from the v4 branch to the main branch so that they are more easily found. The problem with that is that then they wouldn't be in the v4 anymore, so if someone clones that branch and starts working, they won't see the CONTRIBUTING.md. Or one updates the files in both branches, which seems like not the best idea to me.

I would probably advise against this, as in my opinion they should always stay where the main code (and in this case docs) is. There could always be a direct link to it, wherever it is, from the `README.md`.

> I guess this is why most repos have the development branch as the default branch.

Perhaps so x) I my experience, seeing the docs, or the main file structure of the project itself, be on a different branch then the default is quite rare. Seeing the the project files like `CONTRIBUTING.md`, `CODE_OF_CONDUCT.md` and the `LICENCE` away from the main branch is something I have yet to see though.

> As you can see, this is something I'm kind of split about. I also just updated the README to make this structure clearer, linking to the v4 branch directly at the top as well.

I think that's completely fine, there's valid pro and cons for both avenues in my opinion. I just took a look at it and I already think that's better and clearer for future potential contributors.

---

> I actually would find the perfect structure to be a branch called main containing; like it does now the README, and resources like software preference files, in addition to the aforementioned CONTRIBUTING.md / CODE_OF_CONDUCT.md file (and here comes the punchline), a folder called pages containing the markdown files and a folder called images. The second branch would then be called backend and would contain the quartz Node.js. This would nicely separate the content from the website architecture.
>
> The deployment action could then somehow deploy the backend and fetch the pages and images from the main branch.

This is possible, but it in my opinion it would make contributing and maintaining process a lot harder.
To lead the user to these files, it would probably be more advisable to make links to them in the `README.md`.

> The first problem I can think of, however, with this approach, is that building the handbook locally would be a lot more tricky. One would have to modify the Node.js serve command.

My thoughts exactly.

> What I had once also thought about in the past (although I see this comes with far too many downsides) is to have the handbook only contain the content files and have a separate repository that deploys the handbook, fetching the content pages from the handbook repo.

This is also possible, and is arguably better than the modified-serve command option, but again, I feel it would make contributing maintaining quite a bit harder.

---

Cheers 🍻

**Reply by sidney-eliot** [reply]
> If not, would you be interested in checking out these options?

I'm a big fan of Astro's hydration-based HTML rendering approach and have used it for a personal project before.

And yeah, you are correct, that Astro or those other frameworks have ways of dumping HTML or even markdown files into folders that then are arranged on the websites in the way the folder structure is structured. But for the handbook, it was important to me to have a solution that doesn't take much time, so that the bulk of the time spent on the handbook could be put towards actually writing pages.

And seeing as how much hard work Jacky and Ben from the quartz team are putting into that project, it definitely seems like the right call.

---

I'll take a look at those links to projects you posted tomorrow.

**Reply by sugan-reden** [reply]
> ...  But for the handbook, it was important to me to have a solution that doesn't take much time, so that the bulk of the time spent on the handbook could be put towards actually writing pages.
>
> And seeing as how much hard work Jacky and Ben from the quartz team are putting into that project, it definitely seems like the right call.

That is a fair point, zero-to-first-build time is an important thing to consider. I can't really talk for Astro, but the other mentioned options seemed to be pretty fast in my opinion in that regard, as long as you have the required dependencies installed, which aren't much, mostly Node.js and Python/pip.

Quartz is already quite a compelling option to me too, the only things that "bother" me (it's not that much really) is the mini-graph having some wacky text rendering -- at least compared to Obsidian Publish -- and the missing infinite rabbit-hole links from mystmd, which are arguably niche and nit-picky things. 🤷‍♂️ 

---

> I'll take a look at those links to projects you posted tomorrow.

Okie dokie 👍

**Reply by sidney-eliot** [reply]
> infinite rabbithole links from mystmd

Can you explain this or link something?

**Reply by sugan-reden** [reply]
It should be the first feature presented [here](https://mystmd.org/guide#cool-myst-features). 👍

**Reply by sugan-reden** [reply]
Also, I think I may have already mentioned it, but what are your thoughts on the option to export to PDF?

**Reply by sidney-eliot** [reply]
> It should be the first feature presented [here](https://mystmd.org/guide#cool-myst-features). 👍

Oh, I love that feature from Baldur's Gate 3. But it doesn't seem like that would bring much benefit to the handbook, especially as there aren't too many page links currently and unlike Baldur's Gate the user mostly doesn't want to quickly read one sentence from a link and the return to the page they were in. The only place that could be useful for is links to terms in the glossary.

Also, quartz can do internal page link previews for the first level, although it's definitely not very well implemented with the window size not even being pullable.

**Reply by sidney-eliot** [reply]
> Also, I think I may have already mentioned it, but what are your thoughts on the option to export to PDF?

It would be a cool feature, but not reason enough to change frameworks. I have not had an opportunity yet to inform myself about those frameworks you suggested, and will do it some time this week.

**Reply by sugan-reden** [reply]
> > It should be the first feature presented [here](https://mystmd.org/guide#cool-myst-features). 👍
> 
> Oh, I love that feature from Baldur's Gate 3. But it doesn't seem like that would bring much benefit to the handbook, especially as there aren't too many page links currently and unlike Baldur's Gate the user mostly doesn't want to quickly read one sentence from a link and the return to the page they were in. The only place that could be useful for is links to terms in the glossary.

I thought about it some more and I now agree more with your line of thought as the main benefit I saw was indeed for glossary terms. 👍 

> Also, quartz can do internal page link previews for the first level, although it's definitely not very well implemented with the window size not even being pullable.

I agree that it does not feel optimal for me either.
This may be too far-fetched and akin to speculation, but I think for first level viewing I would maybe like something more like [Zen Browser](https://zen-browser.app/) does with its [Zen Glance](https://zen-browser.app/#features) feature.

**Reply by sugan-reden** [reply]
> ... I have not had an opportunity yet to inform myself about those frameworks you suggested, and will do it some time this week.

That's totally fine with me 👍. There's no rush 😄.

**Comment by sidney-eliot** [comment]
Just noticed this:

"It would be a cool feature, but not reason enough to change frameworks. I have not had an opportunity yet to inform myself about those frameworks you suggested, and will do it some time this week."

Was replied to the wrong quote of yours, it should have been:

> Also, I think I may have already mentioned it, but what are your thoughts on the option to export to PDF?

Ironically, you didn't notice as what I accidentally quoted was also fitting

**Reply by sugan-reden** [reply]
Oh damn ^^ That indeed flew right by me 😆.

**Comment by sidney-eliot** [comment]
To even consider switching frameworks, the new framework would have to bring tangible benefits to the table. Switching to a framework where something as basic as page load times all of a sudden feels noticeably worse is not a good value proposition (for this reason I would also discregard [MySTmd ](https://chrisholdgraf.com/blog/2024/mystmd-with-the-blog) even though I do like it). 

Something I also hate is non-proper preloading. The best approach, I think we can all agree, is skeleton screens, and the second best approach being empty pages until the rendering is done (or other sneaky methods like how Vue press brings up a new page with and upwards motion, which hides the load time in the motion and tricks our brain into thinking the load time is shorter due to something happening). As an example, [Shibuya](https://shibuya.lepture.com/customisation/announcement/) loads pages terribly where first one only sees code blocks and internal page links followed by everything else. 

But then again, preloading shouldn't matter anyway, when page load times are instant.

Backend wise, the framework should have one or two centralized files where all the site's CSS can be easily edited. Another thing would be the generation of the folder structure. Anything that requires maintaining a sitemap and other such files would be a downgrade to what quartz currently offers.

And then there's all the other small things quartz offers, like back links and reference links (pages that reference the current page). The graph is obviously also exclusive to quartz, but I'm not sure how essential that feature really is. Quartz's search bar is also better than most of those other frameworks, as it shows the actual page beside it (this is not in the 3d handbook yet, as I still have to update to the newest quartz version).

---

Visually and structure wise, I definitively like some of the sites you suggested, but they should meet most of the requirements I listed above, so if you still think one of these frameworks might be a suitable option, then tell me. 

### Some other options I found as well

Curvenote (MySTmd based):
- https://curvenote.com/docs/publish/authoring-in-myst
- https://phd.row1.ca/

VitePress:
- https://vitepress.dev/guide/sitemap-generation

Gridsome:
- https://gridsome.org/docs/components/

Gitbook:
- https://docs.gitbook.com/
- https://sugoi.gitbook.io/lanraragi/installing-lanraragi/docker
---

So for now, I think sticking with quartz is the best option and if there's something you think needs improving, then one can also directly contribute to quartz and add that feature, jacky is very open when it comes to stuff like that.

**Reply by sugan-reden** [reply]
Performance is indeed a solid point. You are right and I agree that mystmd, and Shibuya have some way to go in that regard, even though I like their look too. However I would argue that, at least in my experience, Quartz has some lacks too (on pc, and bit more on mobile), as well as Curvenote.

Speaking of Curvenote, funnily enough, a good part of the guys behind Curvenote are the same ones working on mystmd (take [Rowan](https://github.com/rowanc1) for example). In my opinion, although I do like the Curvenote almost as much as much as mystmd, it seemed to me to lean a bit too heavily on the scientific side of things (in the sense that has / lacks features that we don't need / need).

> Backend wise, the framework should have one or two centralized files where all the site's CSS can be easily edited. Another thing would be the generation of the folder structure. Anything that requires maintaining a sitemap and other such files would be a downgrade to what quartz currently offers.

I agree with this 👍. I had some CSS tweaks / suggestions I might have liked to make.
I agree for the sitemap generation too. That said, in the case of Sphinx, this is likely fixable with the inclusion of a plugin like [this](https://pypi.org/project/sphinx-sitemap/).

Speaking of Furo, and since you seemed to like VitePress, I might also suggest the [Furo](https://github.com/pradyunsg/furo) theme for Sphinx, which is also a bit more minimalist. It is the theme that Blender uses for the the [Blender Manual](https://docs.blender.org/manual/en/dev/index.html) (see the footer section).

For Gitbook, although I quite like its features as well as visually, I find its pricing model a bit to restrictive. Also, although they claim to be open-source, but not everything is. The performance leaves a bit to be desired too, and self-hosting it is now a pain. This is particularly sad to see, as this project was originally fully open-source, but they have since moved away from it quite a bit.

For Gridsome, I haven't personally had any experience with this, so I can't say much. But from what I'm seeing, this doesn't look too bad, I wouldn't be opposed to this 👍.

Now for VitePress, I do like the aesthetic very much. And I have nothing to say about it's performance. In my opinion, this would definitely be a good option. However, from what I got as per the VuePress docs, [VitePress is a minified, more compressed, opinionated version of VuePress with good defaults](https://vuepress.vuejs.org/guide/introduction.html#vitepress). That said, adding customization and / or plugins would be a bit harder here. To maintain the availability and the ease of customization, I would argue towards using VuePress more (which would make things like adding backlinks not a complete pain, among other things). I would also add that - visually at least - VitePress by default looks the same as [VuePress with the Plume theme](https://theme-plume.vuejs.press/guide/intro/) but sadly the theme's documentation was never made in English it seems. Sitemaps would be available, as well as markdown extensions, and good components.
That said, despite everything I would be down with this 👍.

Sticking with Quartz would also be a perfectly fine option 🙆. Since all options have their pros and cons.

---

I just realized a bit later while digging that VitePress states the ability to customize it with Vite plugins in their docs, which contradicts the VuePress docs a bit 😓. That would make me a bit even more inclined towards VitePress 👍.

**Comment by sugan-reden** [comment, marked as answer]
**For posterity:** As of these enormous commits ([here](https://github.com/sidney-eliot/3d-artists-handbook/commit/d29fcff8ee5ecb8cd862c265bcd66049d2bcd39e) and [here](https://github.com/sidney-eliot/3d-artists-handbook/commit/5e9388987678ae2f81d562f073234c854c03462d)) made by the man himself, @sidney-eliot, the `main` branch has been made the default, along with the whole site structure and content being on it.
If you are interested about finding out more about the reasoning behind past setups that resulted in this change, you can read a bit more from this thread at your convenience, starting from [here](https://github.com/sidney-eliot/3d-artists-handbook/discussions/9#discussioncomment-11556936).

**Comment by sidney-eliot** [comment]
I would like to pull stuff from here into a new discussion regarding frontend / backend frameworks.
